pkg/util/informermanager/federatedinformermanager.go

```diff
@@ -0,0 +1,328 @@
+		informerManagers:            map[string]InformerManager{},
+		informerManagersCancelFuncs: map[string]context.CancelFunc{},
+		queue:                       workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter()),
+		logger:                      klog.LoggerWithName(klog.Background(), "federated-informer-manager"),
+	}
+
+	clusterInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			cluster := obj.(*fedcorev1a1.FederatedCluster)
+			return util.IsClusterJoined(&cluster.Status)
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { manager.enqueue(obj) },
+			UpdateFunc: func(_ interface{}, obj interface{}) { manager.enqueue(obj) },
+			DeleteFunc: func(obj interface{}) { manager.enqueue(obj) },
+		},
+	})
+
+	return manager
+}
+
+func (m *federatedInformerManager) enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		m.logger.Error(err, "Failed to enqueue FederatedCluster")
+		return
+	}
+	m.queue.Add(key)
+}
+
+func (m *federatedInformerManager) worker() {
+	key, shutdown := m.queue.Get()
+	if shutdown {
+		return
+	}
+	defer m.queue.Done(key)
+
+	_, name, err := cache.SplitMetaNamespaceKey(key.(string))
+	if err != nil {
+		m.logger.Error(err, "Failed to process FederatedCluster")
+		return
+	}
+
+	cluster, err := m.clusterInformer.Lister().Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		m.logger.Error(err, "Failed to process FederatedCluster, will retry")
+		m.queue.Add(key)
+		return
+	}
+	if apierrors.IsNotFound(err) || !util.IsClusterJoined(&cluster.Status) {
+		if err := m.processClusterDeletion(name); err != nil {
+			m.logger.Error(err, "Failed to process FederatedCluster, will retry")
+			m.queue.Add(key)
+			return
+		}
+		return
+	}
+
+	err, needReenqueue := m.processCluster(cluster)
+	if err != nil {
+		if needReenqueue {
+			m.logger.Error(err, "Failed to process FederatedCluster, will retry")
+		} else {
+			m.logger.Error(err, "Failed to process FederatedCluster")
+		}
+	}
+	if needReenqueue {
+		m.queue.Add(key)
+	}
+}
+
+func (m *federatedInformerManager) processCluster(cluster *fedcorev1a1.FederatedCluster) (err error, needReenqueue bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	clusterName := cluster.Name
+
+	connectionHash, err := m.clientGetter.ConnectionHash(cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get connection hash for cluster %s: %w", clusterName, err), true
+	}
+	if oldConnectionHash, exists := m.connectionMap[clusterName]; exists {
+		if !bytes.Equal(oldConnectionHash, connectionHash) {
+			// This might occur if a cluster was deleted and recreated with different connection details within a short
+			// period of time and we missed processing the deletion. We simply process the cluster deletion and
+			// reenqueue.
+			// Note: updating of cluster connetion details, however, is still not a supported use case.
```

Comment:
If we enforce the immutability of connection details, can we simply check the UIDs instead of hashes?

Reply:
Same for `InformerManager`

Reply:
Yes to the first question.

For the second question, which part of `InformerManager` are you referring to exactly?